src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    internal class SequencePipeReader : PipeReader
+    {
+        private ReadOnlySequence<byte> _sequence;
+        private bool _isReaderCompleted;
+
+        private int _cancelNext;
+
+        public SequencePipeReader(ReadOnlySequence<byte> sequence)
+        {
+            _sequence = sequence;
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            ThrowIfCompleted();
+
+            // Fast path: did we consume everything?
+            if (consumed.Equals(_sequence.End))
+            {
+                _sequence = ReadOnlySequence<byte>.Empty;
+                return;
+            }
```

Comment:
Is calling AdvanceTo on an already empty sequence the majority use case?  If not, I'd expect this to actually be a deoptimization.

Reply:
This isn't checking if the sequence was already empty when the read started. This is checking if the entire ReadOnlySequence returned by ReadAsync was consumed which is very common.

Reply:
Ah. Ok.